### PR TITLE
WIP sample cleanup.

### PR DESCRIPTION
**Pull request by floooh**
...just FYI, I'm going through the samples, make them work, and do some subjective code cleanup.

I'll let you know when I consider them 'done'.

PS: I would also actually prefer to drastically simplify the build.zig, for instance kick out the WebAssembly stuff (which IMHO is just distracting).

I would also prefer a build process that's more aligned with what's usual for the D ecosystem (e.g. if possible without Zig involvement, or at least use Zig only to build the C sources if D can't do that on its own).

Is there actually such a thing like a standard "D build system"? PS: nvm, it seems to be the same mess as in the C/C++ world :/

fix: #13

### CODE REVIEW

**Comment by floooh** [comment]
PS: I'm also thinking about removing the range helper function alltogether, using D's `.ptr` and `.sizeof` works well enough IMHO, and it's less obscure.

**Comment by kassane** [comment]
> I would also prefer a build process that's more aligned with what's usual for the D ecosystem (e.g. if possible without Zig involvement, or at least use Zig only to build the C sources if D can't do that on its own).
>
> Is there actually such a thing like a standard "D build system"? PS: nvm, it seems to be the same mess as in the C/C++ world :/

Sadly, dub only allows the management of packages (like cargo), but it is not suitable for building systems.
Up to now, it has not been possible (without a workaround) to bind C code to D code in dub.

There's the reggae build-system and meson which offer official support for dlang.

- https://github.com/atilaneves/reggae
- https://mesonbuild.com/D.html

**Comment by floooh** [comment]
Ok, then let's stick with Zig :)

Still not sure about the whole WebAssembly thing though ;)

File: src/examples/debugtext.d [hunk 1]
```diff
@@ -1,7 +1,7 @@
 //------------------------------------------------------------------------------
-//  debugtext_print.d
+//  debugtext.d
 //
-//  Demonstrates formatted printing with sokol.debugtext
+//  Test sokol_debugtext.h
 //------------------------------------------------------------------------------
 module examples.debugtext_print;
```
Comment (kassane) [inline src/examples/debugtext.d:6]: Missing rename `module` name. Affect wasm32 target.

https://github.com/kassane/sokol-d/blob/c20da5b0ad4e3e6d42da84ff92b0349297cead53/build.zig#L557

```bash
zig build run-debugtext -Dtarget=wasm32-emscripten -Doptimize=ReleaseSmall
run-debugtext
└─ run /home/kassane/.cache/zig/p/12203d61e8947f25993a86b4d73e2d91b4c1d96f1aa07c581b7dffba7cb5ca91bc35/upstream/emscripten/emrun
   └─ emcc
      └─ zig build-obj debugtext ReleaseSmall wasm32-emscripten 1 errors
error: unable to check cache: stat file '/home/kassane/sokol-d/zig-cache/o/7ba9215d07edf97684e70f3615ac16da/examples.debugtext.o' failed: FileNotFound
```

Comment (floooh) [reply src/examples/debugtext.d:6]: ah oops, thanks I missed that.

Comment (floooh) [reply src/examples/debugtext.d:6]: I'll do a search and replace, missed a lot of other places (dub.sdl etc)

Comment (floooh) [reply src/examples/debugtext.d]: should be fixed now

**Comment by floooh** [comment]
Hmm, I'm now at a point again where I need debugging to work, and I can't get it to work on macOS. I'll try to continue on Linux tomorrow.

**Comment by floooh** [comment]
Is this necessary? It seems to prevent debugging in vscode on macOS:

https://github.com/kassane/sokol-d/blob/c20da5b0ad4e3e6d42da84ff92b0349297cead53/build.zig#L363-L365

...when removing this could, debugging via CodeLLDB "just works". Should I remove it?

**Comment by kassane** [comment]
> Is this necessary? It seems to prevent debugging in vscode on macOS:

Remove it.

**Comment by floooh** [comment]
I also stumbled again over D setting undefined floats to NaN, I'm now changing the sokol-tools code generator to default-zero-initialize all primitive types.

**Comment by floooh** [comment]
Btw what's the purpose of this Zig file?

https://github.com/kassane/sokol-d/blob/main/src/handmade/math.zig

Does D not have an equivalent of those math functions?

**Comment by floooh** [comment]
Ok, the matrix multiplication in math.d were just confusing rows and columns:

https://github.com/kassane/sokol-d/pull/15/commits/ccf6ab4cbb00afd01bc363844653d005165ca6d8

**Comment by floooh** [comment]
...btw, maybe it's just me but I find the D feature of calling a method when the code looks like it's a field access quite confusing (e.g. `v.len` vs `v.len()`)... I'll change those places to method call syntax.

**Comment by floooh** [comment]
Also... I feel like we should change the sokol-shdc code-generated shader-desc function names from `bla_shader_desc()` to `blaShaderDesc()` to have a consistent function/method naming convention.

**Comment by floooh** [comment]
Ok, all samples fixed except user-data (taking a short break and then I'll wrap it up)

Screenshot 2024-05-12 at 15 50 19

Screenshot 2024-05-12 at 15 51 54

Screenshot 2024-05-12 at 15 50 36

Screenshot 2024-05-12 at 15 50 03

Screenshot 2024-05-12 at 15 55 01


Screenshot 2024-05-12 at 15 50 54

Screenshot 2024-05-12 at 15 51 16

Screenshot 2024-05-12 at 15 51 32

**Comment by floooh** [comment]
Q: is the `cStrTod` required? I don't see is used anywhere and we could remove the entire util.d file if it is not needed (I already removed the asRange helper).

**Comment by kassane** [comment]
Any unused functionality may be removed. `cStrTod`, remove.
If the user needs to convert string, they can use `.ptr` or `to!(T)(mystring)` - need `std.conv : to;`.

**Comment by kassane** [comment]
> Ok, all samples fixed except user-data (taking a short break and then I'll wrap it up)


Awesome. Please don't forget to add your username in the license and dub file author section

**Comment by floooh** [comment]
> Did you get the result in wasm?

I haven't tested wasm so far.

**Comment by floooh** [comment]
What's the output of the `user_data` sample supposed to look like, does this look about right?

```
ExampleUserData(2, [])
ExampleUserData(3, [])
ExampleUserData(4, [])
ExampleUserData(5, [])
ExampleUserData(6, [])
ExampleUserData(7, [])
ExampleUserData(8, [])
ExampleUserData(9, [])
ExampleUserData(10, [])
ExampleUserData(11, [])
ExampleUserData(12, [])
ExampleUserData(13, [13:56])
ExampleUserData(14, [13:56])
ExampleUserData(15, [13:56])
ExampleUserData(16, [13:56])
ExampleUserData(17, [13:56])
ExampleUserData(18, [13:56])
ExampleUserData(19, [13:56])
ExampleUserData(20, [13:56])
ExampleUserData(21, [13:56])
ExampleUserData(22, [13:56])
ExampleUserData(23, [13:56])
ExampleUserData(24, [13:56])
ExampleUserData(25, [13:56])
ExampleUserData(26, [13:56, 26:112])
ExampleUserData(27, [13:56, 26:112])
ExampleUserData(28, [13:56, 26:112])
ExampleUserData(29, [13:56, 26:112])
ExampleUserData(30, [13:56, 26:112])
ExampleUserData(31, [13:56, 26:112])
...
```

**Comment by floooh** [comment]
Another question: when I look into the CI build.yml, the commands `dub build :clear` etc don't work locally here, I'm getting linker errors like (looks like an ARM Mac thing).

Wouldn't it make more sense to run `zig build` in the CI pipeline?

```
sokol-d ➤ dub build :clear
    Starting Performing "debug" build using /Users/floh/bin/dmd2/osx/bin/dmd for x86_64.
    Building sokol-d:clear 0.1.0: building configuration [application]
   Pre-build Running commands
Undefined symbols for architecture x86_64:
  "_main", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture x86_64
Undefined symbols for architecture x86_64:
  "_main", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: /usr/bin/cc failed with status: 1
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: /usr/bin/cc failed with status: 1
Undefined symbols for architecture x86_64:
  "_main", referenced from:
      <initial-undefines>
```

**Comment by floooh** [comment]
Hmm, trying to build with `-Dtarget=wasm32-emscripten` I get all sorts of problems. First, an error about amissing `_d_delThrowable` related to the try-catch block around writeln: 

```
.../sokol-d/src/examples/clear.d(32,11): Error: undefined identifier `_d_delThrowable`
        } catch (Exception) {}
/Users/floh/.cache/zig/p/122006f8e097d5bd627f26d2aa3b397d92b400c786f13396c47afed66fed1e8ef2c1/arsd-webassembly/arsd/webassembly.d(83,5): Error: static assert:  `0` is false
				static assert(0);
    ^
/Users/floh/.cache/zig/p/122006f8e097d5bd627f26d2aa3b397d92b400c786f13396c47afed66fed1e8ef2c1/arsd-webassembly/std/stdio.d(6,6):        instantiated from here: `eval!(ExampleUserData)`
	eval(q{
     ^
/Users/floh/projects/sokol/bindgen/sokol-d/src/examples/user_data.d(41,20):        instantiated from here: `writeln!(ExampleUserData)`
            writeln(*state);
                   ^
          ^
```

...and after that tons of signature mismatch warnings from wasm-ld (haven't seen anything like this yet):

```
wasm-ld: warning: function signature mismatch: sg_alloc_image
>>> defined as (i32) -> void in /Users/floh/projects/sokol/bindgen/sokol-d/zig-cache/o/77f475841dbd57649f2a6e57d3894bd9/saudio.o
>>> defined as () -> i32 in /Users/floh/projects/sokol/bindgen/sokol-d/zig-cache/o/260d164a145bc09fdca683cf35347b6c/libsokol.a(/Users/floh/projects/sokol/bindgen/sokol-d/zig-cache/o/a1560d8d8f7da281f51dd8f60a97a466/sokol_gfx.o)
```
...this goes on for pages and pages... basically for every C function.

**Comment by floooh** [comment]
But TBH, I would probably move WASM support into a later version, and get the bindings without WASM support out ASAP.

**Comment by kassane** [comment]
> What's the output of the `user_data` sample supposed to look like, does this look about right?

This example was based on the rust version. The result obtained was exactly this compared to the rust output.

> Why 'dub' though, wouldn't it make more sense to run zig build in the CI pipeline? The dub file also looks a bit awkward with those file exceptions [...]

Well, I must admit that the dub is really tedious to use and that it's still the biggest cause of unresolved issues in the community.
The dub has been inserted at the moment by [code.dlang.org](https://code.dlang.org) (like [crates.io](https://crates.io))

It doesn't make much sense to keep both tests on the CI/CD knowing that `zig-build` offers better usage options. However, convincing other D-users to use zig only in Dlang projects..

**Comment by floooh** [comment]
I think this specific PR is ready for review/merge now. The shaders are generated with a sokol-shdc from a branch which I'm going to merge now.

The bindings are generated from the 'merge branch' I have in flight here: https://github.com/floooh/sokol/tree/kassane-dlang_bindgen

**Comment by kassane** [comment]
> Hmm, trying to build with -Dtarget=wasm32-emscripten I get all sorts of problems.

`-Doptimize=ReleaseSmall` skip asserts issues.

**Comment by floooh** [comment]
Ah wait, before merging: what about the generated shader function names (e.g. snake-case instead of Pascal case), should I change that in sokol-shdc and in the samples?

**Comment by kassane** [comment]
Wow! dub issue

based on `build.zig` dflag missing on examples:

```bash
$ dub build -b release --compiler=$HOME/zig-bootstrap/out/host/bin/ldc2 :mrt
    Starting Performing "release" build using /home/kassane/zig-bootstrap/out/host/bin/ldc2 for x86_64.
    Building sokol-d:mrt 0.1.0: building configuration [application]
   Pre-build Running commands
src/examples/mrt.d(148,30): Error: function `makeShader` is not callable using argument types `(ShaderDesc)`
src/examples/mrt.d(148,30):        cannot pass rvalue argument `offscreen_shader_desc(queryBackend())` of type `ShaderDesc` to parameter `ref scope ShaderDesc desc`
src/sokol/gfx.d(1365,8):        `sokol.gfx.makeShader(ref scope ShaderDesc desc)` declared here
src/examples/mrt.d(176,30): Error: function `makeShader` is not callable using argument types `(ShaderDesc)`
src/examples/mrt.d(176,30):        cannot pass rvalue argument `fsq_shader_desc(queryBackend())` of type `ShaderDesc` to parameter `ref scope ShaderDesc desc`
src/sokol/gfx.d(1365,8):        `sokol.gfx.makeShader(ref scope ShaderDesc desc)` declared here
src/examples/mrt.d(205,30): Error: function `makeShader` is not callable using argument types `(ShaderDesc)`
src/examples/mrt.d(205,30):        cannot pass rvalue argument `dbg_shader_desc(queryBackend())` of type `ShaderDesc` to parameter `ref scope ShaderDesc desc`
src/sokol/gfx.d(1365,8):        `sokol.gfx.makeShader(ref scope ShaderDesc desc)` declared here
Error /home/kassane/zig-bootstrap/out/host/bin/ldc2 failed with exit code 1.
```

```diff
diff --git a/dub.sdl b/dub.sdl
index 5015679..c4c6736 100644
--- a/dub.sdl
+++ b/dub.sdl
@@ -111,6 +111,7 @@ subPackage {
        targetPath "build"
        sourceFiles "src/examples/mrt.d"
        libs "sokol"
+       dflags "-preview=all"
        lflags "-Lzig-out/lib" platform="posix"
        lflags "/LIBPATH:zig-out/lib" platform="windows"
        excludedSourceFiles "src/examples/sgl_context.d" "src/examples/clear.d" "src/examples/saudio.d" "src/examples/debugtext.d" "src/examples/triangle.d" "src/examples/sgl_points.d" "src/examples/user_data.d" "src/examples/cube.d" "src/examples/blend.d" "src/shaders/*.d"
```

```bash
$ dub build -b release --compiler=$HOME/zig-bootstrap/out/host/bin/ldc2 :mrt
    Starting Performing "release" build using /home/kassane/zig-bootstrap/out/host/bin/ldc2 for x86_64.
    Building sokol-d:mrt 0.1.0: building configuration [application]
   Pre-build Running commands
     Linking sokol-d_mrt
/usr/bin/ld: /home/kassane/.dub/cache/sokol-d/0.1.0/+mrt/build/application-release-peIithuoEG1mqDCbAxtCcQ/sokol-d_mrt.o: in function `init':
time.d:(.text.init+0x1ff): undefined reference to `offscreen_shader_desc'
/usr/bin/ld: time.d:(.text.init+0x411): undefined reference to `fsq_shader_desc'
/usr/bin/ld: time.d:(.text.init+0x56d): undefined reference to `dbg_shader_desc'
collect2: error: ld returned 1 exit status
Error: /usr/bin/cc failed with status: 1
Error /home/kassane/zig-bootstrap/out/host/bin/ldc2 failed with exit code 1.
```

**Comment by floooh** [comment]
When trying to build the WASM version with `-Doptimize=ReleaseSmall`, the build works (still getting all that wasm-ld warning spam), but only the clear sample works, the others "crash" in the browser with:

```
onerror: RuntimeError: Unreachable code should not be executed (evaluating 'Q.get(a)(d,b)')
RuntimeError: Unreachable code should not be executed (evaluating 'Q.get(a)(d,b)')
```

...no idea what's up with that, but could be related to those warnings?

**Comment by floooh** [comment]
Regarding dub, yeah I get similar errors, only the clear sample compiles.

For instance in the triangle sample:

```
dub build :triangle                                                                                                             git:floooh-fixes-1
    Starting Performing "debug" build using ldc2 for aarch64, arm_hardfloat.
    Building sokol-d:triangle 0.1.0: building configuration [application]
   Pre-build Running commands
src/examples/triangle.d(47,30): Error: function `sokol.gfx.makeShader(ref scope ShaderDesc desc)` is not callable using argument types `(ShaderDesc)`
src/examples/triangle.d(47,30):        cannot pass rvalue argument `triangle_shader_desc(queryBackend())` of type `ShaderDesc` to parameter `ref scope ShaderDesc desc`
```

**Comment by kassane** [comment]
> Ah wait, before merging: what about the generated shader function names (e.g. snake-case instead of Pascal case), should I change that in sokol-shdc and in the samples?

If you can maintain a sokol style by default, it helps or see: https://dlang.org/dstyle.html

**Comment by kassane** [comment]
The main issue with dub is that it requires reading file by file as opposed to autodetect-local `-i` flag like zig is doing.

**Comment by floooh** [comment]
Regarding this error:

```
...cannot pass rvalue argument...
```

I wonder if that's because I got rid of temporary variables, and instead nested the calls needed to create a shader. But then why did it work in build.zig?

Here's my toolchain version btw (via `ldc2 --version`

```
LDC - the LLVM D compiler (1.37.0):
  based on DMD v2.107.1 and LLVM 17.0.6
  built with LDC - the LLVM D compiler (1.37.0)
  Default target: arm64-apple-darwin23.4.0
  Host CPU: apple-m1
  http://dlang.org - http://wiki.dlang.org/LDC
```

**Comment by kassane** [comment]
> ...cannot pass rvalue argument...

remove `-preview=all` - experimental features D standard. Like clang `-fexperimental-library`.
DIP 1000 (safe rules) affects rvalues.

**Comment by floooh** [comment]
Ok, when I change this code:

```
sg.makeShader(shd.triangle_shader_desc(sg.queryBackend()));
```

...so that it uses an intermediate variable like this:

```
auto shd_desc = shd.triangle_shader_desc(sg.queryBackend());
sg.makeShader(shd_desc);
```
...it works. Blargh, that sucks :D

I also get a linker error, but I guess that's because I moved the formerly platform-specific embedded shader code in the triangle sample into a `.glsl` shader.

**Comment by floooh** [comment]
...do I need to declare an imported source file (e.g. `import shd = shaders.triangle;`) also as a linker dependency somewhere in the build file?

**Comment by kassane** [comment]
> ...do I need to declare an imported source file (e.g. `import shd = shaders.triangle;`) also as a linker dependency somewhere in the build file?

`-i[=<pattern>]                                 - Include imported modules in the compilation`

**Comment by floooh** [comment]
Another problem I encounter with dub :/

```
src/examples/cube.d(128,26): Error: cannot take address of local `vsParams` in `@safe` function `frame`
```

...for code like this:

```
shd.VsParams vsParams = { mvp: computeMvp(state.rx, state.ry) };
sg.Range r = { ptr: &vsParams, size: vsParams.sizeof };
```

...why suddenly all those errors? Is this a "D 1.0" vs "D 2.0" thing?

**Comment by kassane** [comment]
betwen zig and dub have `dflags "-preview=all"`?

**Comment by floooh** [comment]
Since `-preview=all` makes a lot more sense than vanilla D (to me at least lol), shouldn't we enable it in the dub build too?

**Comment by kassane** [comment]
> Another problem I encounter with dub :/

Fix dub here

```bash
LANG=C dub build -b release --compiler=$HOME/zig-bootstrap/out/host/bin/ldc2 :cube
    Starting Performing "release" build using /home/kassane/zig-bootstrap/out/host/bin/ldc2 for x86_64.
    Building sokol-d:cube 0.1.0: building configuration [application]
   Pre-build Running commands
     Linking sokol-d_cube
kassane:~/sokol-d $ LANG=C dub build -b release --compiler=$HOME/zig-bootstrap/out/host/bin/ldc2 :mrt
    Starting Performing "release" build using /home/kassane/zig-bootstrap/out/host/bin/ldc2 for x86_64.
    Building sokol-d:mrt 0.1.0: building configuration [application]
   Pre-build Running commands
     Linking sokol-d_mrt
kassane:~/sokol-d $ git diff
diff --git a/dub.sdl b/dub.sdl
index 5015679..66cb6a2 100644
--- a/dub.sdl
+++ b/dub.sdl
@@ -34,6 +34,7 @@ subPackage {
        targetPath "build"
        sourceFiles "src/examples/clear.d"
        libs "sokol"
+       dflags "-preview=all" "-i"
        lflags "-Lzig-out/lib" platform="posix"
        lflags "/LIBPATH:zig-out/lib" platform="windows"
        excludedSourceFiles "src/examples/sgl_context.d" "src/examples/triangle.d" "src/examples/sgl_points.d" "src/examples/saudio.d" "src/examples/debugtext.d" "src/examples/mrt.d" "src/examples/user_data.d" "src/examples/cube.d" "src/examples/blend.d" "src/shaders/*.d"
```

**Comment by kassane** [comment]
> a lot more sense than vanilla D (to me at least lol), shouldn't we enable it in the dub build too?

When we use zig only, it is responsible for ldc2 execution.
However, with dub enabled, zig is only responsible for libsokol (clang) only.

**Comment by floooh** [comment]
Ok, the dub builds work with those additional flags.

But next problem: it looks like the dub builds want to use a libsokol dynamic link library is this intended?

On macOS I get this when I want to run one of the dub-generated exes:

```
sokol-d ➤ build/sokol-d_triangle
dyld[9896]: Library not loaded: @rpath/libsokol.dylib
```

**Comment by kassane** [comment]
ref.: #13 

> The Windows target using the sokol dynamic library does not work. Only works in static library using zig build by default.

**Comment by floooh** [comment]
Can't we use static linking instead?

**Comment by kassane** [comment]
> Can't we use static linking instead?

need add some sokol config on dub. On zig sokol share with ldc2Config.

**Comment by floooh** [comment]
Hmm ok, I committed the dub.sdl changes (-preview=all -i), and then let me also quickly change the naming convention for the create-shader function... I guess that's all I can do for now.

**Comment by kassane** [comment]
> Can't we use static linking instead?

On Linux e.g.:
```bash
LANG=C dub build -b release --compiler=$HOME/zig-bootstrap/out/host/bin/ldc2 :cube
    Starting Performing "release" build using /home/kassane/zig-bootstrap/out/host/bin/ldc2 for x86_64.
    Building sokol-d:cube 0.1.0: building configuration [application]
   Pre-build Running commands
     Linking sokol-d_cube
/usr/bin/ld: zig-out/lib/libsokol.a(/home/kassane/sokol-d/zig-cache/o/a72e20601cb44fb59375c61f7cb39bec/sokol_app.o): in function `_sapp_linux_run':
sokol_app.c:(.text+0x4bd): undefined reference to `XInitThreads'
/usr/bin/ld: sokol_app.c:(.text+0x4c2): undefined reference to `XrmInitialize'
/usr/bin/ld: sokol_app.c:(.text+0x4c9): undefined reference to `XOpenDisplay'
/usr/bin/ld: sokol_app.c:(.text+0x548): undefined reference to `XkbSetDetectableAutoRepeat'
/usr/bin/ld: sokol_app.c:(.text+0x554): undefined reference to `XResourceManagerString'
/usr/bin/ld: sokol_app.c:(.text+0x561): undefined reference to `XrmGetStringDatabase'
/usr/bin/ld: sokol_app.c:(.text+0x598): undefined reference to `XrmGetResource'
/usr/bin/ld: sokol_app.c:(.text+0x5c5): undefined reference to `XrmDestroyDatabase'
/usr/bin/ld: sokol_app.c:(.text+0x634): undefined reference to `XInternAtom'
/usr/bin/ld: sokol_app.c:(.text+0x650): undefined reference to `XInternAtom'
/usr/bin/ld: sokol_app.c:(.text+0x66c): undefined reference to `XInternAtom'
/usr/bin/ld: sokol_app.c:(.text+0x688): undefined reference to `XInternAtom'
/usr/bin/ld: sokol_app.c:(.text+0x6a4): undefined reference to `XInternAtom'
/usr/bin/ld: zig-out/lib/libsokol.a(/home/kassane/sokol-d/zig-cache/o/a72e20601cb44fb59375c61f7cb39bec/sokol_app.o):sokol_app.c:(.text+0x6c0): more undefined references to `XInternAtom' follow
/usr/bin/ld: zig-out/lib/libsokol.a(/home/kassane/sokol-d/zig-cache/o/a72e20601cb44fb59375c61f7cb39bec/sokol_app.o): in function `_sapp_linux_run':
```

**Comment by kassane** [comment]
dub static-lib support superset this PR. Add in TODO/Roadmap

**Comment by floooh** [comment]
On Linux:

> undefined reference to `XInternAtom'

...those look like the system DLL dependencies are missing, e.g. (in the sokol-zig build-zig):

```zig
        if (options.use_x11) {
            lib.linkSystemLibrary("X11");
            lib.linkSystemLibrary("Xi");
            lib.linkSystemLibrary("Xcursor");
        }
```

**Comment by floooh** [comment]
I fixed the shader-desc function names btw, so the examples source code should be fine now. I'll merge the sokol-shdc PR now and create binaries, so at least that part would be done.

**Comment by kassane** [comment]
> On Linux:
> 
> > undefined reference to `XInternAtom'
> 
> ...those look like the system DLL dependencies are missing, e.g. (in the sokol-zig build-zig):
> 
> ```zig
>         if (options.use_x11) {
>             lib.linkSystemLibrary("X11");
>             lib.linkSystemLibrary("Xi");
>             lib.linkSystemLibrary("Xcursor");
>         }
> ```

Add in dub
```
lflags "-Lzig-out/lib" platform="posix"
lflags "/LIBPATH:zig-out/lib" platform="windows"
lflags "/LIBPATH:zig-out/lib" platform="osx"
```

**Comment by floooh** [comment]
Ok, the sokol-shdc branch has been merged: https://github.com/floooh/sokol-tools/pull/128

The executables in https://github.com/floooh/sokol-tools-bin will be uptodate when this CI pipeline goes green:

https://github.com/floooh/sokol-tools/actions/runs/9053315598

I'll log off for today then :)

**Comment by floooh** [comment]
PS: I would prefer to merge this PR first, and then you can figure out how to fix the dub files. Since I don't have any knowledge how this works it's better if you take care of that directly ;)

In general I would prefer if we could put out the first version of the bindings soon-ish even if the extra things (like WASM support or building via dub) doesn't work yet. What should work though is building via build.zig on Windows, macOS and Linux.

**Comment by kassane** [comment]
done for merge?

**Comment by floooh** [comment]
> done for merge?

Yep, I think that's all :)

**Comment by kassane** [comment]
Thanks for help.

**Comment by floooh** [comment]
Btw, if you need to re-build the bindings via `gen_all.py` keep in mind that I have this inflight branch here were I dabbled with the `scope` keyword in function args (it's only a minimal change though and if needed I can discard this inflight-branch again):

https://github.com/floooh/sokol/tree/kassane-dlang_bindgen

...hopefully no more changes will be needed to the codegen python scripts before the release (e.g. ideally only changes to the sokol-d repo).

**Comment by kassane** [comment]
@floooh, discard my PR and merge your branch.

Dub fixed for static-lib on v0.1.1

**Pull request merged**